### PR TITLE
[Security] Allow in memory user to have extra fields

### LINKED ISSUE #40091: Allow extra fields on memory users config.
**Description**  
Actually, if you use memory provider, you can only define few attributes (username, password, roles).
It will be nice, if you can define custom attributes.

**Example**
security.yml
```
security:
    providers:
        app_provider:
            memory:
                users:
                    adrien:  { password: pwd, roles: ['ROLE_USER'], extra: {gender: m} }

```
Somewhere.php
`  $user->getExtraFields()['gender'];`
Class `Symfony\Component\Security\Core\User\User` already has $extraFields attribute.

### PR DESCRIPTION
**Pull request by noniagriconomie**
| Q             | A
| ------------- | ---
| Branch?       | 5.x 
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #40091
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/14928

---

```yaml
# config/packages/security.yaml
security:
    providers:
        users_in_memory:
            memory:
                users:
                    aa: {password: 'aa', roles: 'ROLE_USER'}
                    bb: {password: 'bb', roles: ['ROLE_ADMIN', 'ROLE_USER'], extra_fields: {'name': 'John', 'age': 77}}
```

```php
// $user -> Symfony\Component\Security\Core\User\User
$user->getExtraFields()['age']; // 77
```

### CODE REVIEW

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/UserProvider/InMemoryFactory.php [hunk 1]
```diff
@@ -57,6 +57,9 @@ public function addConfiguration(NodeDefinition $node)
                                 ->beforeNormalization()->ifString()->then(function ($v) { return preg_split('/\s*,\s*/', $v); })->end()
                                 ->prototype('scalar')->end()
                             ->end()
+                            ->arrayNode('extraFields')
```
Comment (noniagriconomie) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/UserProvider/InMemoryFactory.php:60]: WIP part, I need to read doc and test how to format this node

File: src/Symfony/Bundle/SecurityBundle/CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+5.3.0
+-----
```
Comment (OskarStark) [inline src/Symfony/Bundle/SecurityBundle/CHANGELOG.md]:
```suggestion
 
5.3
---
```

File: src/Symfony/Bundle/SecurityBundle/CHANGELOG.md [hunk 3]
```diff
@@ -1,6 +1,11 @@
...
+
+ * Added a new `extraFields` in `memory` provider to allow user to have extra fields
```
Comment (OskarStark) [inline src/Symfony/Bundle/SecurityBundle/CHANGELOG.md]:
```suggestion
 * Add a new `extraFields` in `memory` provider to allow user to have extra fields
```

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php [hunk 4]
```diff
@@ -301,7 +301,7 @@ private function addProvidersSection(ArrayNodeDefinition $rootNode)
                             'memory' => [
                                 'users' => [
                                     'foo' => ['password' => 'foo', 'roles' => 'ROLE_USER'],
-                                    'bar' => ['password' => 'bar', 'roles' => '[ROLE_USER, ROLE_ADMIN]'],
+                                    'bar' => ['password' => 'bar', 'roles' => '[ROLE_USER, ROLE_ADMIN]', 'extraFields' => ['age' => '77']],
```
Comment (OskarStark) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: Are could be an integer

File: src/Symfony/Bundle/SecurityBundle/CHANGELOG.md (hunk 2 above)
Comment (noniagriconomie) [reply src/Symfony/Bundle/SecurityBundle/CHANGELOG.md]: all lines below have 3 digits @OskarStark

File: src/Symfony/Bundle/SecurityBundle/CHANGELOG.md (hunk 3 above)
Comment (noniagriconomie) [reply src/Symfony/Bundle/SecurityBundle/CHANGELOG.md]: changed

File: src/Symfony/Bundle/SecurityBundle/CHANGELOG.md (hunk 2 above)
Comment (OskarStark) [reply src/Symfony/Bundle/SecurityBundle/CHANGELOG.md]: Fabien documented how to write `CHANGELOG.md` files here, please keep a look and adjust it accordingly.

https://github.com/symfony/symfony-docs/pull/14830

Thanks 🙏

Comment (noniagriconomie) [reply src/Symfony/Bundle/SecurityBundle/CHANGELOG.md]: I will have a look at it 👍

File: src/Symfony/Bundle/SecurityBundle/CHANGELOG.md [hunk 5]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+5.3
```
Comment (chalasr) [inline src/Symfony/Bundle/SecurityBundle/CHANGELOG.md]: should be `5.3.0`

File: src/Symfony/Component/Security/Core/User/InMemoryUserProvider.php [hunk 6]
```diff
@@ -79,15 +80,15 @@ public function refreshUser(UserInterface $user)
...
+        return new User($storedUser->getUsername(), $storedUser->getPassword(), $storedUser->getRoles(), $storedUser->isEnabled(), $storedUser->isAccountNonExpired(), $storedUser->isCredentialsNonExpired() && $storedUser->getPassword() === $user->getPassword(), $storedUser->isAccountNonLocked(), $storedUser->getExtraFields());
     }
 
     /**
      * {@inheritdoc}
      */
     public function supportsClass(string $class)
     {
-        return 'Symfony\Component\Security\Core\User\User' === $class;
+        return User::class === $class;
```
Comment (chalasr) [inline src/Symfony/Component/Security/Core/User/InMemoryUserProvider.php]: unrelated change potentially causing merge conflicts, should be reverted

**Review by chalasr — CHANGES_REQUESTED** [review]
A [functional test](https://github.com/symfony/symfony/tree/5.x/src/Symfony/Bundle/SecurityBundle/Tests/Functional) would be nice :)

File: src/Symfony/Bundle/SecurityBundle/CHANGELOG.md (hunk 5 above)
Comment (noniagriconomie) [reply src/Symfony/Bundle/SecurityBundle/CHANGELOG.md]: @chalasr please have a look at https://github.com/symfony/symfony/pull/40097#discussion_r570579611
@OskarStark points me to "new changelog" way :)

Comment (chalasr) [reply src/Symfony/Bundle/SecurityBundle/CHANGELOG.md]: oh, I must have missed it, sorry!

**Comment by stof** [comment]
Does this really fit in the architecture of the new system where this User class is considered for deprecation ?

**Comment by ro0NL** [comment]
given the current situation, i figured the mapping was missing from an implementation pov, hence im 👍🏻

however i do think we should rename it to `InMemoryUser` eventually

`LdapUser` provides the same feature set, let alone user entites in the wild providing the full app feature set ... then this seems reasonable :)

**Comment by wouterj** [comment]
I personally only see a use-case for the in-memory provider in simple demo applications. Do we really want to advocate using it in real applications? If the answer is no, then I'm in favor of not adding more lines that we need to maintain.

(btw, I would not say that the User is up for deprecation, but hopefully the credential related methods are. Any security system needs to have some sort of user/principal)

**Comment by ro0NL** [comment]
https://symfony.com/doc/current/security/user_provider.html#memory-user-provider

> It’s not recommended to use this provider in real applications because of its limitations and how difficult it is to manage users. 

we fix some limitation here. Managing YAML is debatable 😁 

>  It may be useful in application prototypes and for limited applications that don’t store users in databases.

So it is useful for limited (non-prototype) apps actually?

IMHO it's just a persistence layer, and if in-memory fits ... it fits :)

**Comment by wouterj** [comment]
It's not a "real" persistence layer, as it requires to use Symfony's User. Btw, it's very easy to implement your own in-memory user provider that uses a custom User class with all properties it needs:

```php
class MyInMemoryUserProvider implements UserProvider
{
    private const USERS = [
        'wouter' => [...],
        // ...
    ];

    public function loadUserByUsername(string $username)
    {
        if (!$user = self::USERS[$username] ?? false) {
            throw new UsernameNotFoundException();
        }

        return new AppUser($user['username'], $user['password'], ...);
    }

    public function refreshUser(UserInterface $user)
    {
        return $this->loadUserByUsername($user->getUsername());
    }

    public function supportsClass(string $class)
    {
        return is_a($class, AppUser::class);
    }
}
```

So I'm honestly leaning more and more towards not adding more properties to the `User` class.

**Comment by Nyholm** [comment]
> I personally only see a use-case for the in-memory provider in simple demo applications. 

For context, 
I've used in-memory provider on apps that only have one or two API users. Think of an internal microservice with a HTTP API.

**Comment by noniagriconomie** [comment]
Hello, i am not aware of the deprecation of this class, where can i find more?

Also, as the prop already exists, i think it is more « finishing » the implementation than adding a new one
I do use very much this feature of in memory provider for apps where i do need only one or very few users (api, unitary admin) and having this extrafields could help avoid adding user related data in a « hacky » way

**Comment by chalasr** [comment]
We don't need to deprecate this class and the corresponding provider, only renaming it (already possible) and make it follow the rework of the UserInterface _when it's time to_. 
This makes configurable a property that already exists on something that has no on-going rework and has been proven useful, nothing fancy here.
There is no reason to "feature freeze" this code area, I think we should relax and just let the work happen as it comes.

**Comment by chalasr** [comment]
Rebase needed. Don't hesitate to tell us if you need help to finish this PR @noniagriconomie

**Comment by noniagriconomie** [comment]
@chalasr ok I will rebase asap
for the help, I may ask for some around the [functional test](https://github.com/symfony/symfony/pull/40097#pullrequestreview-584396706) you asked previously
are there any specific test file(s) in this folder I can take inspiration from? thank you

**Comment by noniagriconomie** [comment]
@chalasr rebase done and a test added
review appreciated :)

File: src/Symfony/Component/Security/Core/Tests/User/InMemoryUserProviderTest.php [hunk 7]
```diff
@@ -25,6 +25,7 @@ public function testConstructor()
         $user = $provider->loadUserByUsername('fabien');
         $this->assertEquals('foo', $user->getPassword());
         $this->assertEquals(['ROLE_USER'], $user->getRoles());
+        $this->assertEquals(['name' => 'John', 'age' => 77, 'salary' => 1234.56], $user->getExtraFields());
```
Comment (OskarStark) [inline src/Symfony/Component/Security/Core/Tests/User/InMemoryUserProviderTest.php:28]: assertSame ?

Same for the others in this file

**Review by OskarStark — APPROVED** [review]

File: src/Symfony/Component/Security/Core/Tests/User/InMemoryUserProviderTest.php (hunk 7 above)
Comment (noniagriconomie) [reply src/Symfony/Component/Security/Core/Tests/User/InMemoryUserProviderTest.php:28]: I do not know if it is related to this PR to change the method name

and for this new added line I just duplicate the above line to keep things similar

if asked/decided I can handle it inside this PR of course :)

**Comment by fabpot** [comment]
I agree with @wouterj and I don't think we want to add this feature. I understand the use case and I have the same on some small applications. But as already mentioned, there are simple ways to achieve what you want without support in the framework. Even if adding it in core in not a lot of code, I think that this sends the wrong signal.

**Comment by OskarStark** [comment]
@noniagriconomie would you then send the proposed code snippet by Wouter as a docs PR to help others achieve this more easily? Thanks

**Comment by wouterj** [comment]
For reference: the `extraFields` property in `User` was added specially for LDAP: https://github.com/symfony/symfony/pull/31532 (back then, the specialized `LdapUser` didn't exists yet). It may even make sense to deprecate that functionality in the main `User` imho.

**Comment by noniagriconomie** [comment]
hey, ok understood even if I really thought this was a nice addition
lets close for now, I will see if a doc snippet could be usefull
